backend/clickhouse/traces.go

```diff
@@ -381,9 +362,70 @@ func (client *Client) ReadTrace(ctx context.Context, projectID int, traceID stri
 		return a.Timestamp.Before(b.Timestamp)
 	})
 
-	span.Finish(rows.Err())
+	if rows.Err() != nil {
+		return nil, rows.Err()
+	}
 
-	return traces, rows.Err()
+	return traces, nil
+}
+
+func (client *Client) ReadTrace(ctx context.Context, projectID int, traceID string) ([]*modelInputs.Trace, error) {
+	sb := sqlbuilder.NewSelectBuilder()
+	var err error
+	var args []interface{}
+
+	sb.From(TracesByIdTable).
+		Select("Timestamp, UUID, TraceId, SpanId, ParentSpanId, ProjectId, SecureSessionId, TraceState, SpanName, SpanKind, Duration, ServiceName, ServiceVersion, Environment, HasErrors, TraceAttributes, StatusCode, StatusMessage").
+		Where(sb.Equal("ProjectId", projectID)).
+		Where(sb.Equal("TraceId", traceID))
+
+	sql, args := sb.BuildWithFlavor(sqlbuilder.ClickHouse)
+
+	span, _ := util.StartSpanFromContext(ctx, "traces", util.ResourceName("ReadTrace"))
+
+	rows, err := client.conn.Query(ctx, sql, args...)
+	if err != nil {
+		span.Finish(err)
+		return nil, err
+	}
+
+	traces, err := getTracesFromRows(rows)
+	if err != nil {
+		span.Finish(err)
+		return nil, err
+	}
+
+	span.Finish()
+
+	if len(traces) > 0 {
+		return traces, nil
+	}
+
+	span, _ = util.StartSpanFromContext(ctx, "traces", util.ResourceName("ReadTraceFallback"))
+
+	sb.From(TracesTable).
+		Select("Timestamp, UUID, TraceId, SpanId, ParentSpanId, ProjectId, SecureSessionId, TraceState, SpanName, SpanKind, Duration, ServiceName, ServiceVersion, Environment, HasErrors, TraceAttributes, StatusCode, StatusMessage").
+		Where(sb.Equal("ProjectId", projectID)).
+		Where(sb.Equal("TraceId", traceID)).
+		Where(sb.GE("Timestamp", time.Now().Add(-5*time.Minute))).
+		Where(sb.LE("Timestamp", time.Now()))
```

Comment:
👍 this limits us from searching too many records when this should only happen to very recent records